llm_unlearn_ucl/unlearn_harm.py

```diff
@@ -54,64 +57,64 @@ def set_seed(seed_num: int) -> None:
-    device: torch.device,
-    compute_for_answer_only: bool = True,
-) -> List[float]:
-    # Compute the average of min-K% Prob values for the entire batch
-    # TODO: should we do this on just 2 element batch or entire dataset?
-    #
-    # NOTE: Need to properly unpack the batch, compute logits for each batch and
-    # properly mask the output! only compute min K on the part we are checking if is
-    # Unlearned/member in pretraining: output
-    with torch.no_grad():
-        # TODO: verify: in run.py for mink, calculatePerplexity func (49), they also pass labels=input_ids. WHy?
-        # TODO: should we: feeed full sentance (question + answer), and then mask out to compute probabilities,
-        # Or Mask out input, only get logits for the answer and compute probabilities on those
-        # NOTE: For now, we feed full question, as we are unlearning B given A, so we want individual token
-        # probabilities of B, given A (but we don't care about token probabilities of A - the question)
-        outputs = model(
-            batch["input_ids"].to(device),
-            attention_mask=batch["attention_mask"].to(device),
-        )
-        # OR, something along thelines of
-        # outputs = model(batch["input_ids"][batch["start_locs"]:], attention_mask=batch["attention_mask"])
-
-    logits = outputs.logits
-    probabilities = torch.nn.functional.log_softmax(logits, dim=-1)
-    no_of_sentences = batch["input_ids"].shape[0]
-    # If computing for both question and answer, need to set start_locs to 0s!
-    if compute_for_answer_only:
-        assert batch.get("start_locs") != None, (
-            "Compute Min-k % prob: Requested computation only for answer, "
-            "but the batch does not contain start_locs inside tokenised question+answer pairs!"
-        )
-    else:
-        # start locs by default just after <s> starting token!
-        batch["start_locs"] = [torch.IntTensor([1]) for _ in range(no_of_sentences)]
-
-    # Compute for each sentence in the batch
-    pred_mink = []
-    for s_idx in range(no_of_sentences):
-        # extract prob for each token in the unlearned answer given all previous tokens
-        input_ids_sentence = batch["input_ids"][s_idx]  # [1:]
-        all_prob = []
-        for i in range(batch["start_locs"][s_idx].item(), len(input_ids_sentence)):
-            token_id = input_ids_sentence[i]
-            # i is 1 ahead of the actual token, as we want the probability of that token given all previous tokens
-            probability = probabilities[s_idx, i - 1, token_id].item()
-            all_prob.append(probability)
-        # Get top-K % probs and compute their mean (it was log_softmax, so top k% is actually mink% prob)
-        k_length = int(len(all_prob) * K)
-        topk_prob = np.sort(all_prob)[:k_length]
-        pred_mink.append(-np.mean(topk_prob).item())
-
-    # All mean MIN-K% prob in: pred_mink. For now, return all
-    return pred_mink
+# def compute_mink_prob(
```

Comment:
Please remove, we no longer need it. In the worst case we will look back in git history :)

🎶For the first time in **history**
It's gonna start raining men (Start raining men)! 🎶